opds_import.py

```diff
@@ -925,50 +879,143 @@ def __init__(self, _db, feed_url, default_data_source, import_class,
+                    new_data = True
+                    break
 
-    def follow_one_link(self, link, start):
-        self.log.info("Following next link: %s, cutoff=%s", link, start)
-        response = self._get(link)
+                if  updated >= cutoff:
+                    # This book has been updated.
+                    new_data = True
+                    break
 
-        imported, messages, next_links = self.importer.import_from_feed(
-            response.content, even_if_no_author=True, cutoff_date=start,
+            if record and record.exception:
+                # We tried to import this book before and it failed. The book hasn't been updated,
+                # so it probably still won't work. Don't try again.
+                continue
+
+            # There's no record of a failure for this book. Has it been imported from this source already?
+            existing_edition = get_one(
+                self._db, Edition,
+                primary_identifier=identifier,
+                data_source=data_source,
+            )
+            if not existing_edition:
+                new_data = True
+                break
+
+        return new_data
+
+    def follow_one_link(self, link, cutoff_date=None, do_get=None):
+        self.log.info("Following next link: %s, cutoff=%s", link, cutoff_date)
+        get = do_get or self._get
+        status_code, content_type, feed = get(link, None)
+
+        new_data = self.check_for_new_data(feed, cutoff_date=cutoff_date)
+
+        if new_data:
+            # There's something new on this page, so we need to check the next page as well.
+            next_links = self.importer.extract_next_links(feed)
+            return next_links, feed
+        else:
+            # There's nothing new, so we don't need to import this feed or check the next page.
+            return [], None
+
+    def import_one_feed(self, feed, start):
+        imported_editions, pools, works, messages = self.importer.import_from_feed(
+            feed, even_if_no_author=True, cutoff_date=start,
             immediately_presentation_ready = self.immediately_presentation_ready
         )
-        self._db.commit()
+
+        data_source = DataSource.lookup(self._db, self.importer.data_source_name)
         
-        if len(imported) == 0:
-            # We did not end up importing a single book on this page.
-            # There's no need to keep going.
-            self.log.info(
-                "Saw a full page with no new or updated books. Stopping."
+        # Create CoverageRecords for the successful imports.
+        for edition in imported_editions:
+            CoverageRecord.add_for(
+                edition, data_source, CoverageRecord.IMPORT_OPERATION
             )
-            return []
-        else:
-            return next_links
 
+        # Create CoverageRecords for the failures.
+        for urn, message in messages.items():
+            if message.status_code == 500:
```

Comment:
I could be wrong but I think you're not distinguishing between the local 500 errors which get turned into StatusMessages when there's an exception, and 500 errors from the remote side. Local 500 errors are permanent because they require a change to the code, but remote 500 errors should be treated as transient, because they could be fixed at any time.

What if `messages` was allowed to contain `CoverageFailure` objects as well as `StatusMessage` objects? This would let you differentiate between local and remote errors.